clients/windows/src/input/file_drop.rs

```diff
@@ -0,0 +1,172 @@
+    ) -> Result<()> {
+        // indicates to the drop source that they don't need to delete the source data (unlike DROPEFFECT_MOVE)
+        // "If DoDragDrop returns DROPEFFECT_MOVE, delete the source data from the source document immediately. No other return value from DoDragDrop has any effect on a drop source."
+        // https://learn.microsoft.com/en-us/cpp/mfc/drag-and-drop-ole?view=msvc-170
+        unsafe { *effect = DROPEFFECT_COPY };
+
+        (self.handler)(Message::DragEnter { object: object.cloned(), state, point: *point });
+        Ok(())
+    }
+
+    fn DragOver(
+        &self, state: MODIFIERKEYS_FLAGS, point: &POINTL, _: *mut DROPEFFECT,
+    ) -> Result<()> {
+        (self.handler)(Message::DragOver { state, point: *point });
+        Ok(())
+    }
+
+    fn DragLeave(&self) -> Result<()> {
+        (self.handler)(Message::DragLeave);
+        Ok(())
+    }
+
+    fn Drop(
+        &self, object: Option<&IDataObject>, state: MODIFIERKEYS_FLAGS, point: &POINTL,
+        _: *mut DROPEFFECT,
+    ) -> Result<()> {
+        (self.handler)(Message::Drop { object: object.cloned(), state, point: *point });
+        Ok(())
+    }
+}
+
+pub fn handle_drop(app: &mut WgpuLockbook, object: Option<IDataObject>) -> bool {
+    if let Some(object) = object {
+        let format_enumerator: IEnumFORMATETC = unsafe {
+            object
+                .EnumFormatEtc(DATADIR_GET.0 as _)
+                .expect("enumerate drop formats")
+        };
+        let mut rgelt = [FORMATETC::default(); 1];
+        loop {
+            let mut fetched: u32 = 0;
+            if unsafe { format_enumerator.Next(&mut rgelt, Some(&mut fetched as _)) }.is_err() {
+                break;
+            }
+            if fetched == 0 {
+                break;
+            }
+
+            let format = CLIPBOARD_FORMAT(rgelt[0].cfFormat);
+            let mut format_name = [0u16; 1000];
+            let is_predefined_format = format_str(format).is_some();
+            let is_registered_format =
+                unsafe { GetClipboardFormatNameW(format.0 as _, &mut format_name) != 0 };
+            if !is_predefined_format && !is_registered_format {
+                continue;
+            }
+
+            let stgm = unsafe { object.GetData(&rgelt[0]) }.expect("get drop data");
+
+            let tymed = TYMED(stgm.tymed as _);
+            if tymed_str(tymed).is_none() {
+                continue;
+            }
+
+            if tymed == TYMED_HGLOBAL {
+                let hglobal = unsafe { stgm.u.hGlobal };
+
+                // for unknown reasons, if I don't cast the HGLOBAL to an HDROP and query the file count, the next call to object.GetData fails
+                // (this applies even if the format isn't CF_HDROP)
+                let hdrop = HDROP(unsafe { std::mem::transmute(hglobal) });
+                let file_count = unsafe { DragQueryFileW(hdrop, 0xFFFFFFFF, None) };
+
+                if format == CF_HDROP {
```

Comment:
a couple missing variants of these guys mean we don't support most types of drops, but we do support dropping a file dragged from file explorer so we're doing alright